File: MinimumCostToFillGivenWeightInABag/MinimumCostToFillGivenWeightInABag.java
/*
https://www.geeksforgeeks.org/problems/minimum-cost-to-fill-given-weight-in-a-bag1956/1?utm_source=geeksforgeeks&utm_medium=newui_home&utm_campaign=potd
*/

//{ Driver Code Starts
import java.io.*;
import java.util.*;

class IntArray {
    public static int[] input(BufferedReader br, int n) throws IOException {
        String[] s = br.readLine().trim().split(" ");
        int[] a = new int[n];
        for (int i = 0; i < n; i++) a[i] = Integer.parseInt(s[i]);

        return a;
    }

    public static void print(int[] a) {
        for (int e : a) System.out.print(e + " ");
        System.out.println();
    }

    public static void print(ArrayList<Integer> a) {
        for (int e : a) System.out.print(e + " ");
        System.out.println();
    }
}

class GFG {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t;
        t = Integer.parseInt(br.readLine());
        while (t-- > 0) {

            int n;
            n = Integer.parseInt(br.readLine());

            int w;
            w = Integer.parseInt(br.readLine());

            int[] cost = IntArray.input(br, n);

            Solution obj = new Solution();
            int res = obj.minimumCost(n, w, cost);

            System.out.println(res);
        }
    }
}

// } Driver Code Ends



class Solution {
    
    public static long depthFirstSearch_forArrays(int totalNumberOfPackets, int weightRequired, int[] cost, int[] weightOfThePackets, long[][] memoization){
        if(weightRequired==0) return 0;
        if(totalNumberOfPackets<=0||weightRequired<0) return Integer.MAX_VALUE;
        if(memoization[totalNumberOfPackets][weightRequired]!=-1) return memoization[totalNumberOfPackets][weightRequired];
        long minimumCostToBuyExactlyWKgOranges=Integer.MAX_VALUE;
        long resultAfterPickingCurrentIndexPacketForFinalCost=Integer.MAX_VALUE;
        long resultAfterNotPickingCurrentIndexPacketForFinalCost=Integer.MAX_VALUE;
        if(cost[totalNumberOfPackets-1]!=-1){
            resultAfterNotPickingCurrentIndexPacketForFinalCost=depthFirstSearch_forArrays(totalNumberOfPackets-1, weightRequired, cost, weightOfThePackets, memoization);
            minimumCostToBuyExactlyWKgOranges=Math.min(minimumCostToBuyExactlyWKgOranges, Math.min(resultAfterPickingCurrentIndexPacketForFinalCost, resultAfterNotPickingCurrentIndexPacketForFinalCost));
        }
        if(weightOfThePackets[totalNumberOfPackets-1]<=weightRequired){
            resultAfterPickingCurrentIndexPacketForFinalCost=cost[totalNumberOfPackets-1]+depthFirstSearch_forArrays(totalNumberOfPackets, weightRequired-weightOfThePackets[totalNumberOfPackets-1], cost, weightOfThePackets, memoization);
            resultAfterNotPickingCurrentIndexPacketForFinalCost=depthFirstSearch_forArrays(totalNumberOfPackets-1, weightRequired, cost, weightOfThePackets, memoization);
            minimumCostToBuyExactlyWKgOranges=Math.min(minimumCostToBuyExactlyWKgOranges, Math.min(resultAfterPickingCurrentIndexPacketForFinalCost, resultAfterNotPickingCurrentIndexPacketForFinalCost));
        } else{
            resultAfterNotPickingCurrentIndexPacketForFinalCost=depthFirstSearch_forArrays(totalNumberOfPackets-1, weightRequired, cost, weightOfThePackets, memoization);
            minimumCostToBuyExactlyWKgOranges=Math.min(minimumCostToBuyExactlyWKgOranges, Math.min(resultAfterPickingCurrentIndexPacketForFinalCost, resultAfterNotPickingCurrentIndexPacketForFinalCost));
        }
        memoization[totalNumberOfPackets][weightRequired]=minimumCostToBuyExactlyWKgOranges;
        return memoization[totalNumberOfPackets][weightRequired];
    }
    
    public static int minimumCost(int n, int w, int[] cost) {
        int[] weightsOfThePackets=new int[n];
        for(int packetIIndex=0; packetIIndex<n; ++packetIIndex){
            weightsOfThePackets[packetIIndex]=packetIIndex+1;
        }
        long[][] memoization=new long[n+1][w+1];
        for(int packetIIndex=0; packetIIndex<=n; ++packetIIndex){
            for(int weightIndexOfPacketI=0; weightIndexOfPacketI<=w; ++weightIndexOfPacketI){
                memoization[packetIIndex][weightIndexOfPacketI]=-1L;
            }
        }
        long minimumCostToBuyExactlyWKgOranges=depthFirstSearch_forArrays(n, w, cost, weightsOfThePackets, memoization);
        if(minimumCostToBuyExactlyWKgOranges>=Integer.MAX_VALUE){
            return -1;
        }
        return (int)minimumCostToBuyExactlyWKgOranges;
    }
}

File: MinimumCostToFillGivenWeightInABag/MinimumCostToFillGivenWeightInABag.cpp
/*
https://www.geeksforgeeks.org/problems/minimum-cost-to-fill-given-weight-in-a-bag1956/1?utm_source=geeksforgeeks&utm_medium=newui_home&utm_campaign=potd
*/

//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

class Array {
  public:
    template <class T>
    static void input(vector<T> &A, int n) {
        for (int i = 0; i < n; i++) {
            scanf("%d ", &A[i]);
        }
    }

    template <class T>
    static void print(vector<T> &A) {
        for (int i = 0; i < A.size(); i++) {
            cout << A[i] << " ";
        }
        cout << endl;
    }
};


// } Driver Code Ends

class Solution {
  public:
  
    long long dfs(int n, int w, vector<int>&cost, vector<int>&weight, vector<vector<int>>&memo){
        if(w==0) return 0;
        if(n<=0||w<0) return INT_MAX;
        if(memo[n][w]!=-1) return memo[n][w];
        long long res=INT_MAX, pk=INT_MAX, ntpk=INT_MAX;
        if(cost[n-1]!=-1){
            ntpk=dfs(n-1, w, cost, weight, memo);
            res=min(res, min(pk, ntpk));
        }
        if(weight[n-1]<=w){
            pk=cost[n-1]+dfs(n, w-weight[n-1], cost, weight, memo);
            ntpk=dfs(n-1, w, cost, weight, memo);
            res=min(res, min(pk, ntpk));
        } else{
            ntpk=dfs(n-1, w, cost, weight, memo);
            res=min(res, min(pk, ntpk));
        }
        return memo[n][w]=res;
    }
  
    int minimumCost(int n, int w, vector<int> &cost) {
        vector<int>weight(n);
        for(int i=0; i<n; ++i){
            weight[i]=i+1;
        }
        vector<vector<int>>memo(n+1, vector<int>(w+1, -1));
        int fnres=dfs(n, w, cost, weight, memo);
        return fnres>=INT_MAX?-1:fnres;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    scanf("%d ", &t);
    while (t--) {

        int n;
        scanf("%d", &n);

        int w;
        scanf("%d", &w);

        vector<int> cost(n);
        Array::input(cost, n);

        Solution obj;
        int res = obj.minimumCost(n, w, cost);

        cout << res << endl;
    }
}

// } Driver Code Ends